Event retention documentation (#1602)
OctopusDeploy/docs: docs/security/users-and-teams/auditing.md

For team members to collaborate in the deployment of software, there needs to be trust and accountability. Octopus Deploy captures audit information whenever significant events happen in the system.

> :::warning

Reviewer: ```suggestion
:::hint
```

Author: Argh, I meant to accept this suggestion... I'll do it manually 🙂

> The Event Retention functionality is being rolled out in stages for our Cloud customers. So you may get access to that functionality at a later stage.

Reviewer: ```suggestion
The [Audit Retention functionality](#manage-archived-audit-events) was introduced in **Octopus 2022.3** and is being rolled out in stages for our Cloud customers first. It will be made available to Self-hosted customers at a later stage.
```

> The [Audit Retention functionality](#manage-archived-audit-events) was introduced in **Octopus 2022.3** and is being rolled out in stages for our Cloud customers first. It will be made available to Self-hosted customers at a later stage.

Reviewer: ```suggestion
The [Audit Retention functionality](#manage-archived-audit-events) was introduced in **Octopus 2022.4** and is being rolled out in stages for our Cloud customers first. It will be made available to Self-hosted customers at a later stage.
```

Reviewer: According to the latest information I saw, this would now be introduced in 2022.4

- Octopus **does** capture the details of every mutating action (create/edit/delete) including who initiated the action.
- Octopus **does not** capture login and logout events for specific user accounts.
- Octopus **does not** capture when data is read, however certain sensitive actions like downloading a certificate with its private key is captured.

> - Entries older than 90 days (default, configurable up to 365 days) are archived and deleted from the database. As such they are not queryable through the user interface. 

Reviewer: ```suggestion
- From **Octopus 2022.3** and later, entries older than 90 days (default, configurable up to 365 days) are archived and deleted from the database. As such, they are not queryable through the user interface. 
```

Reviewer: This is only a suggestion. Not sure how it works with any feature flag usage for self-hosted customers

Reviewer 2: @harrisonmeister we mention a version at the top as well as the fact that not all customers will see that. So I think it is sufficient and we shouldn't repeat it again.

Reviewer 3: @harrisonmeister as @irynasemenova mentioned, we do list the version at the top and as you noted, with feature toggles it's going to be difficult to predict exactly when on-prem are going to see this. I'm comfortable with the callout at the top of the dock. 

@sburmanoctopus @irynasemenova this did get me thinking about the wording here. I wonder if we could tweak it to something like the follow, just to be explicit -
`Entries older than 90 days (default, configurable up to 365 days) are archived and deleted from the database. Archived events are not queryable through the user interface.`

Author: Sounds good. Done!

Reviewer: Works for me :+1:

To grant a user access to audit logs you can make use of a built-in User Role that contains **EventView**. All project related user roles contain it. **EventView** can also be scoped to narrow down which audit information a user can see, for example, it can be restricted to specific Projects or Environments. Learn more about [managing users and teams](/docs/security/users-and-teams/index.md).

> Audit entries older than the configured retention period (defaults to 90 days, configurable up to 365 days) are archived and can be access via the overflow menu (three dots) in the top right corner of the audit log page ("Manage archived events").

Reviewer: ```suggestion
Audit entries older than the configured retention period (defaults to 90 days, configurable up to 365 days) are archived and can be accessed via the overflow menu (`...`) in the top right corner of the audit page by selecting the **Manage archived events** option.
```

In **Octopus 2019.1** we removed **AuditView** in an effort to simplify permissions so only **EventView** is now required.

### Modifying and deleting audit logs is prevented

Octopus actively prevents modifying or deleting audit logs within the configured retention period via its API. That being said, a user with the appropriate permissions to the `Events` table in your Octopus SQL Database could modify or delete records in that table. If you are concerned about this kind of tampering you should configure the permissions to the `Events` table in your Octopus SQL Database appropriately.

> Entries older than the retention period can be deleted by users with the appropriate permissions (typically `Octopus Manager`). An audit log entry will be created each time an archived event file is deleted. You can prevent archived files from being deleted by setting the configured folder to be read-only at the filesystem level.

Reviewer: How does this work in practice? By limiting a folder to be read-only to prevent deletion of files, would this not also prevent the ability of Octopus Server to write the files to the archive folder too?

Author: I see what you mean. 

I think what this is trying to say is that other users should have their access level set to read-only.

Maybe an alternative could be (replacing 'You can prevent archived files...'):

`Archived files are saved at a filesystem level. So any user that has the appropriate permissions could view or delete these files. If this is a concern, you should restrict the permissions to access the configured folder appropriately`

@irynasemenova, what do you think?

Reviewer 2: > would this not also prevent the ability of Octopus Server to write the files to the archive folder too

Is this question on the User's permission? Or Octopus application permission? @harrisonmeister 

> `Archived files are saved at a filesystem level. So any user that has the appropriate permissions could view or delete these files. If this is a concern, you should restrict the permissions to access the configured folder appropriately`

Yeah, this does seem clearer @sburmanoctopus

Author: 👍 I've updated the text 🙂

Reviewer: The question was on the Octopus application's permissions as it wasn't immediately clear. The new text is :+1:

### Sensitive values in audit logs

If you make a change to a sensitive value in Octopus, you will notice we write an audit log showing the fact the sensitive value changed. The value we show in the audit log is simply **an indicator the value has changed**. This is **not** the unencrypted/raw value. This is **not** even the encrypted value.

We take the sensitive value and hash it using an irreversible hash algorithm. We then encrypt that hash with a new, unique, non-deterministic salt. We use this irreversible value as an indicator that the sensitive value actually changed in some way.

> ### Manage archived audit logs

Reviewer: ```suggestion
### Manage archived audit logs {#manage-archived-audit-events}
```

> Audit log entries can require significant amount of database space to store, which gradually degrades overall system performance. For this reason Octopus Server applies a retention policy to automatically archive audit log entries older than the configured number of days and remove them from the database. The retention period can be configured via **{{Configuration, Settings, EventRetention}}**. The location of the archived audit log files can be changed via **{{Configuration, Settings, ServerFolders}}**.

Reviewer: ```suggestion
:::hint
[Audit Retention functionality](#manage-archived-audit-events) was introduced in **Octopus 2022.3** and is being rolled out in stages for our Cloud customers first. It will be made available to Self-hosted customers at a later stage.
:::

Audit log entries can require a significant amount of database space to store, degrading overall system performance. For this reason, Octopus Server applies a retention policy to automatically archive audit log entries older than the configured number of days and remove them from the database. The retention period can be configured via **{{Configuration, Settings, EventRetention}}**. The location of the archived audit log files can be changed via **{{Configuration, Settings, ServerFolders}}**.
```

> Periodically, Octopus will apply the retention policy to the existing entries and store them as [JSONL](https://jsonlines.org/) files, grouped as a single file per each day (for example, `events-2019-01-01.jsonl`). These files can be accessed through a separate page available via the overflow menu in the top right corner of the main audit log page.

Reviewer: ```suggestion
Periodically, Octopus will apply the retention policy to the existing entries and store them as [JSONL](https://jsonlines.org/) files, grouped as a single file for each day (for example, `events-2019-01-01.jsonl`). These files can be accessed through a separate page available via the overflow menu (`...`) in the top right corner of the audit page.
```

Users with appropriate permissions (typically `Octopus Manager`) can download or delete the archived files. The downloaded files are intended to be imported into a datalake for querying and analysis.

:::warning

> Deleting the archived files will permanently erase the audit entries.

Reviewer: ```suggestion
**Take care deleting archived files**
Deleting the archived files will permanently erase the audit entries.
```

:::

> The archived files can also be accessed via the API endpoints `/api/event/archives` and `/api/event/archives/{filename}`.

Reviewer: ```suggestion
The archived files can also be accessed via the Octopus REST API endpoints `/api/event/archives` and `/api/event/archives/{filename}`.
```

Entries older than the retention period can be deleted by users with the appropriate permissions (typically `Octopus Manager`). An audit log entry will be created each time an archived event file is deleted. Archived files are saved at a filesystem level. So any user that has the appropriate permissions could view or delete these files. If this is a concern, you should restrict the permissions to access the configured folder appropriately.

### Sensitive values in audit logs

If you make a change to a sensitive value in Octopus, you will notice we write an audit log showing the fact the sensitive value changed. The value we show in the audit log is simply **an indicator the value has changed**. This is **not** the unencrypted/raw value. This is **not** even the encrypted value.

We take the sensitive value and hash it using an irreversible hash algorithm. We then encrypt that hash with a new, unique, non-deterministic salt. We use this irreversible value as an indicator that the sensitive value actually changed in some way.

### Manage archived audit logs {#manage-archived-audit-events}

:::hint

> Audit Retention functionality was introduced in **Octopus 2022.3** and is being rolled out in stages for our Cloud customers first. It will be made available to Self-hosted customers at a later stage.

Reviewer: ```suggestion
Audit Retention functionality was introduced in **Octopus 2022.4** and is being rolled out in stages for our Cloud customers first. It will be made available to Self-hosted customers at a later stage.
```

Reviewer: According to the latest information I saw, this would now be introduced in 2022.4

Reviewer 4: We're rolling out to cloud customers in 2022.3 but it will be available for on-prem in 2022.4. Does that mean this should be 2022.4 or should we change the wording?

Reviewer: We could opt to change the wording to:

```suggestion
Audit Retention functionality was introduced in **Octopus 2022.3** and is being rolled out in stages for our Cloud customers first. It will be made available to Self-hosted customers in **Octopus 2022.4**.
```

But the wording is accurate, it might be confusing if anyone wonders what "later stage" means.
